File: room-annotation-processor/src/org/yuyan/room/annotation/processor/helper/AnnotationProcessorHelper.java
package org.yuyan.room.annotation.processor.helper;

import com.squareup.javapoet.*;
import javax.annotation.processing.RoundEnvironment;
import javax.lang.model.element.ExecutableElement;
import javax.lang.model.element.TypeElement;
import java.util.Set;

public class AnnotationProcessorHelper {
    public static final String DB_IMPL_SUFFIX = "_Impl";
    public static final String FIELD_INSTANCE_SUFFIX = "_Instance";

    public static boolean processable(Set<? extends TypeElement> annotations, RoundEnvironment roundEnv) {
        if (roundEnv == null || roundEnv.processingOver()) {
            return false;
        }
        if (annotations == null || annotations.isEmpty()) {
            return false;
        }
        return true;
    }



    public static MethodSpec.Builder formMethodBuilder(ExecutableElement element, boolean isOverride) {
        String methodName = element.getSimpleName().toString();
        MethodSpec.Builder methodBuilder = MethodSpec.methodBuilder(methodName);

        if (isOverride) {
            ClassName override = ClassName.get("java.lang", "Override");
            methodBuilder.addAnnotation(override);
        }

        element.getModifiers().forEach(modifier -> {
            switch (modifier) {
                case PUBLIC:
                case PROTECTED:
                    methodBuilder.addModifiers(modifier);
                case PRIVATE:
                case DEFAULT:
                case STATIC:
                case FINAL:
                case ABSTRACT:
                    break;
            }
        });

        element.getParameters().forEach(variableElement -> {
            String paramName = variableElement.getSimpleName().toString();
            TypeName clsType = TypeName.get(variableElement.asType());
            ParameterSpec.Builder paramBuilder = ParameterSpec.builder(clsType, paramName);
            methodBuilder.addParameter(paramBuilder.build());
        });

        TypeName typeName = TypeName.get(element.getReturnType());
        methodBuilder.returns(typeName);
        return methodBuilder;
    }
}

File: room-annotation-processor/src/org/yuyan/room/annotation/processor/AnnotationProcessorHelper.kt
package org.yuyan.room.annotation.processor

import com.squareup.kotlinpoet.*
import org.yuyan.room.base.RoomDatabase
import javax.annotation.processing.RoundEnvironment
import javax.lang.model.element.Element
import javax.lang.model.element.ExecutableElement
import javax.lang.model.element.TypeElement
import kotlin.reflect.KClass


fun ifProcess(annotations: MutableSet<out TypeElement>?, roundEnv: RoundEnvironment?): Boolean {
    when (roundEnv?.processingOver()) {
        true -> return false
    }
    when (annotations.isNullOrEmpty()) {
        true -> return false
    }
    return true
}

fun findAnnotatedTypeElementByName(clsName: String, inputElements: MutableSet<out Element>?): HashSet<TypeElement>{
    val annotatedTypeElements = HashSet<TypeElement>()
    inputElements?.forEach { element ->
        val mirrors = element.annotationMirrors
        mirrors.forEach { annotationMirror ->
            when (annotationMirror.annotationType.toString()) {
                clsName -> {
                    annotatedTypeElements.add(element = element as TypeElement)
                }
            }
        }
    }
    return annotatedTypeElements
}

fun formClassBuilder(element: TypeElement): TypeSpec.Builder {
    val pkgName: String = element.enclosingElement.toString()
    val clsName: String = element.simpleName.toString()
    val clsImplName = ClassName(
        pkgName, "${clsName}${RoomDatabase.DB_IMPL_SUFFIX}"
    )
    return TypeSpec.classBuilder(clsImplName)
        .addSuperinterface(ClassName(pkgName, clsName))
}

fun formMethodBuilder(element: ExecutableElement): FunSpec.Builder{
    val methodBuilder = FunSpec.builder(element.simpleName.toString())

    methodBuilder.addModifiers(KModifier.OVERRIDE)
    /*element.modifiers.forEach {
        methodBuilder.addModifiers(KModifier.valueOf(it.name))
    }*/
    element.parameters.forEach {
        methodBuilder.addParameter(
            ParameterSpec.builder(
                it.simpleName.toString()
                , it.asType().asTypeName()
            ).build()
        )
    }
    methodBuilder.returns(element.returnType.asTypeName())
    return methodBuilder
}